packages/delegation-core/src/caveats/allowedCalldata.ts

```diff
@@ -0,0 +1,74 @@
+import { bytesToHex, remove0x, type BytesLike } from '@metamask/utils';
+
+import {
+  defaultOptions,
+  prepareResult,
+  type EncodingOptions,
+  type ResultValue,
+} from '../returns';
+import type { Hex } from '../types';
+import { toHexString } from '../utils';
+
+/**
+ * Terms for configuring an AllowedCalldata caveat.
+ */
+export type AllowedCalldataTerms = {
+  startIndex: number;
+  value: BytesLike;
+};
+
+/**
+ * Creates terms for an AllowedCalldata caveat that ensures the provided execution calldata
+ * matches the expected calldata at the specified index.
+ *
+ * @param terms - The terms for the AllowedCalldata caveat.
+ * @param encodingOptions - The encoding options for the result.
+ * @returns The terms as the calldata itself.
+ * @throws Error if the `calldata` is invalid.
+ */
+export function createAllowedCalldataTerms(
+  terms: AllowedCalldataTerms,
+  encodingOptions?: EncodingOptions<'hex'>,
+): Hex;
+export function createAllowedCalldataTerms(
+  terms: AllowedCalldataTerms,
+  encodingOptions: EncodingOptions<'bytes'>,
+): Uint8Array;
+/**
+ * Creates terms for an AllowedCalldata caveat that ensures the provided execution calldata
+ * matches the expected calldata at the specified index.
+ * @param terms - The terms for the AllowedCalldata caveat.
+ * @param encodingOptions - The encoding options for the result.
+ * @returns The terms as the calldata itself.
+ * @throws Error if the `calldata` is invalid.
+ */
+export function createAllowedCalldataTerms(
+  terms: AllowedCalldataTerms,
+  encodingOptions: EncodingOptions<ResultValue> = defaultOptions,
+): Hex | Uint8Array {
+  const { startIndex, value } = terms;
+
+  if (startIndex < 0) {
+    throw new Error('Invalid startIndex: must be zero or positive');
+  }
+
+  if (!Number.isInteger(startIndex)) {
+    throw new Error('Invalid startIndex: must be a whole number');
+  }
+
+  let unprefixedValue: string;
+
+  if (typeof value === 'string') {
```

Comment:
could we use `bytesLikeToHex` here?

Reply:
I think we can probably improve this pattern. There's also `isStrictHexString` exported from @metamask/utils

`bytesLikeToHex` is intended as a result formatter (it's exported from `returns.ts`), so doesn't perform any input validation. 

I'll open a backlog item to review all of the terms builders and align the type parsing patterns.